File: backend/app/db/models/alert.py
import uuid
from datetime import datetime

from sqlalchemy import Boolean, DateTime, Float, ForeignKey, String, Uuid, func
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class Alert(Base):
    """A user-defined rule evaluated on a schedule by the Celery beat task."""

    __tablename__ = "alerts"

    id: Mapped[uuid.UUID] = mapped_column(Uuid, primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("users.id", ondelete="CASCADE"), index=True
    )
    ticker: Mapped[str] = mapped_column(String(20), index=True)
    # price_above | price_below | pct_change_above | risk_above
    alert_type: Mapped[str] = mapped_column(String(30))
    threshold: Mapped[float] = mapped_column(Float)
    active: Mapped[bool] = mapped_column(Boolean, default=True, index=True)
    last_triggered_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())


class AlertEvent(Base):
    """A record of an alert firing — the triggered-alert history."""

    __tablename__ = "alert_events"

    id: Mapped[uuid.UUID] = mapped_column(Uuid, primary_key=True, default=uuid.uuid4)
    alert_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("alerts.id", ondelete="CASCADE"), index=True
    )
    triggered_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    payload: Mapped[dict] = mapped_column(JSONB, default=dict)  # value, threshold, message
    delivered: Mapped[bool] = mapped_column(Boolean, default=False)

File: frontend/types/alert.ts
export type AlertType = "price_above" | "price_below" | "pct_change_above" | "risk_above";

export const ALERT_TYPE_LABELS: Record<AlertType, string> = {
  price_above: "Price above",
  price_below: "Price below",
  pct_change_above: "Daily % move above",
  risk_above: "Risk score above",
};

export interface Alert {
  id: string;
  ticker: string;
  alert_type: AlertType;
  threshold: number;
  active: boolean;
  last_triggered_at: string | null;
  created_at: string | null;
}

export interface AlertEvent {
  id: string;
  alert_id: string;
  ticker: string;
  alert_type: string;
  triggered_at: string | null;
  payload: { value?: number; threshold?: number; message?: string };
}
